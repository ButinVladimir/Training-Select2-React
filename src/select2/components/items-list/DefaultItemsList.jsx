import React, { Fragment } from 'react';
import PropTypes from 'prop-types';
import ListItem from './ListItem';
import './style.css';

const convertItem = (item, onChange) => ({
  key: item.value,
  value: item.value,
  name: item.name,
  checked: item.checked,
  onChange,
});

export default function DefaultItemsList({
  items,
  onChange,
}) {
  const convertedItems = items
    .map(item => (
      <ListItem
        {...convertItem(item, onChange)}
      />
    ));

  return (
    <Fragment>
      {
        convertedItems.length > 0 && (
        <ul className="select2-item-list">
          {convertedItems}
        </ul>)
      }
      {
        convertedItems.length === 0 && (
        <div className="items-not-found">No matching items found</div>)
      }
    </Fragment>
  );
}

DefaultItemsList.propTypes = {
  items: PropTypes.arrayOf(PropTypes.shape({
    value: PropTypes.any.isRequired,
    name: PropTypes.string.isRequired,
  })).isRequired,
  onChange: PropTypes.func.isRequired,
};
