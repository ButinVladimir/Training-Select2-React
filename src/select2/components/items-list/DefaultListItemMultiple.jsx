import React from 'react';
import PropTypes from 'prop-types';

export default function DefaultListItemMultiple({
  value,
  name,
  checked,
  onChange,
}) {
  const id = `select2-${value}`;

  return (
    <li className={checked ? 'checked' : ''}>
      <label htmlFor={id}>
        <input id={id} type="checkbox" checked={checked} onChange={() => onChange(value, name)} />
        {name}
      </label>
    </li>
  );
}

DefaultListItemMultiple.propTypes = {
  value: PropTypes.any.isRequired,
  name: PropTypes.string.isRequired,
  checked: PropTypes.bool.isRequired,
  onChange: PropTypes.func.isRequired,
};
