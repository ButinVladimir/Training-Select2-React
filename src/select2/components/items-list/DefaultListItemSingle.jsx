import React from 'react';
import PropTypes from 'prop-types';

export default function DefaultListItemSingle({
  value,
  name,
  checked,
  onChange,
}) {
  return (
    <li className={checked ? 'checked' : ''}>
      <button type="button" onClick={() => onChange(value, name)}>
        {name}
      </button>
    </li>
  );
}

DefaultListItemSingle.propTypes = {
  value: PropTypes.any.isRequired,
  name: PropTypes.string.isRequired,
  checked: PropTypes.bool.isRequired,
  onChange: PropTypes.func.isRequired,
};
