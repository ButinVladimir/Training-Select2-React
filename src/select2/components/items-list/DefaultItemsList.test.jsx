import React from 'react';
import renderer from 'react-test-renderer';
import DefaultItemsList from './DefaultItemsList';
import DefaultListItemSingle from './DefaultListItemSingle';
import DefaultListItemMultiple from './DefaultListItemMultiple';
import ListItemValue from '../../helpers/list-item-value';

describe('DefaultItemsList', () => {
  test('renders without crashing with empty list and single selection', () => {
    const mockFn = jest.fn();
    const component = renderer.create(<DefaultItemsList
      ListItem={DefaultListItemSingle}
      items={[]}
      onChange={mockFn}
    />);

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
    expect(mockFn).not.toHaveBeenCalled();
  });

  test('renders without crashing with empty list and multiple selection', () => {
    const mockFn = jest.fn();
    const component = renderer.create(<DefaultItemsList
      ListItem={DefaultListItemMultiple}
      items={[]}
      onChange={mockFn}
    />);

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
    expect(mockFn).not.toHaveBeenCalled();
  });

  test('renders without crashing with some items in list and single selection', () => {
    const mockFn = jest.fn();
    const component = renderer.create(<DefaultItemsList
      ListItem={DefaultListItemSingle}
      items={[
        new ListItemValue('name 1', 'value 1', false),
        new ListItemValue('name 2', 'value 2', true),
        new ListItemValue('name 3', 'value 3', false),
      ]}
      onChange={mockFn}
    />);

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
    expect(mockFn).not.toHaveBeenCalled();
  });

  test('renders without crashing with some items in list and multiple selection', () => {
    const mockFn = jest.fn();
    const component = renderer.create(<DefaultItemsList
      ListItem={DefaultListItemMultiple}
      items={[
        new ListItemValue('name 1', 'value 1', false),
        new ListItemValue('name 2', 'value 2', true),
        new ListItemValue('name 3', 'value 3', true),
      ]}
      onChange={mockFn}
    />);

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
    expect(mockFn).not.toHaveBeenCalled();
  });
});
