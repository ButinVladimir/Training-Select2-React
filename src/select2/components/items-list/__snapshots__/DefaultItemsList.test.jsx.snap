// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DefaultItemsList renders without crashing with empty list and multiple selection 1`] = `
<div
  className="items-not-found"
>
  No matching items found
</div>
`;

exports[`DefaultItemsList renders without crashing with empty list and single selection 1`] = `
<div
  className="items-not-found"
>
  No matching items found
</div>
`;

exports[`DefaultItemsList renders without crashing with some items in list and multiple selection 1`] = `
<ul
  className="select2-item-list"
>
  <li
    className=""
  >
    <label
      htmlFor="select2-value 1"
    >
      <input
        checked={false}
        id="select2-value 1"
        onChange={[Function]}
        type="checkbox"
      />
      name 1
    </label>
  </li>
  <li
    className="checked"
  >
    <label
      htmlFor="select2-value 2"
    >
      <input
        checked={true}
        id="select2-value 2"
        onChange={[Function]}
        type="checkbox"
      />
      name 2
    </label>
  </li>
  <li
    className="checked"
  >
    <label
      htmlFor="select2-value 3"
    >
      <input
        checked={true}
        id="select2-value 3"
        onChange={[Function]}
        type="checkbox"
      />
      name 3
    </label>
  </li>
</ul>
`;

exports[`DefaultItemsList renders without crashing with some items in list and single selection 1`] = `
<ul
  className="select2-item-list"
>
  <li
    className=""
  >
    <button
      onClick={[Function]}
      type="button"
    >
      name 1
    </button>
  </li>
  <li
    className="checked"
  >
    <button
      onClick={[Function]}
      type="button"
    >
      name 2
    </button>
  </li>
  <li
    className=""
  >
    <button
      onClick={[Function]}
      type="button"
    >
      name 3
    </button>
  </li>
</ul>
`;
